### PR TITLE
13.0 mig web widget text markdown

### PR DESCRIPTION
**Pull request by KKamaa**
Migration of web_widget_text_markdown to v 13.0

### CODE REVIEW

File: web_widget_text_markdown/__manifest__.py [hunk 1]
```diff
@@ -3,29 +3,22 @@
 #
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 {
-    'name': 'Web Widget Text Markdown',
-    'version': '11.0.1.0.0',
+    "name": "Web Widget Text Markdown",
+    "version": "13.0.1.0.0",
```
Comment (thomaspaulb) [inline web_widget_text_markdown/__manifest__.py:7]: Minor nitpick, but the OCA way is to _first_ do the black/isort/prettier in a separate commit, and then the actual migration to 13.0 in a separate commit, which makes it easier to review (people will just have to look at the last commit). Could you still separate the two?

File: web_widget_text_markdown/__manifest__.py [hunk 2]
```diff
@@ -3,29 +3,22 @@
...
-    ],
-    "qweb": [
-        "static/src/xml/bootstrap_markdown.xml",
-    ],
-    'installable': True,
-    'auto_install': False,
-    'application': False
+    "Komit, "
+    "Sudokeys, "
+    "Kevin, "
```
Comment (thomaspaulb) [inline web_widget_text_markdown/__manifest__.py]: Here should be `"Sunflower IT, "`

File: web_widget_text_markdown/README.rst [hunk 3]
```diff
@@ -73,18 +73,7 @@ Authors
...
-
-* Nicolas Jeudy <nicolas@sudokeys.com>
-* Nguyen Tan Phuc <phuc.nt@komit-consulting.com>
-* Alexandre Díaz <dev@redneboa.es>
-
-Other credits
-~~~~~~~~~~~~~
-
-* Komit https://komit-consulting.com
+* Kevin
```
Comment (thomaspaulb) [inline web_widget_text_markdown/README.rst]: You should leave all the previous Contributors there, and add yourself underneath.

It's actually better/easier to use the OCA 'readme generator' for this, `oca-gen-addon-readme`.

So you'd

1. Look at the OCA template module and copy some example files from there such as [CONTRIBUTORS.rst](https://github.com/OCA/maintainer-tools/tree/master/template/module/readme)
2. Fill it in with the correct guys and copy it [here](https://github.com/KKamaa/web/tree/0ec75cc1a65b58dbb0b264d24e5a38d812f2eb32/web_widget_text_markdown/readme)
3. Install `oca_gen_addon_readme` on your system. [Guide](https://github.com/OCA/maintainer-tools/blob/master/README.md#installation) or just:
```
$ git clone git@github.com:OCA/maintainer-tools.git
$ pip install --user ./maintainer-tools
```

4. Run it [like this](https://github.com/OCA/maintainer-tools/blob/master/README.md#readme-generator)

File: web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js [hunk 4]
```diff
@@ -2,75 +2,78 @@
...
+                        data: this.data,
+                        arch: this.arch,
+                        res_id: 1,
+                        viewOptions: {mode: "edit"},
+                    }).then(function(form) {
+                        _.defer(function() {
+                            assert.strictEqual(form.$(".md-editor").length, 1);
+                            form.destroy();
+                            done();
+                        });
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js:73]: Example in point - I can review this better if the black/isort stuff is in a separate commit :-)

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

**Comment by thomaspaulb** [comment]
@KKamaa In order to actually test this on the instance, can you migrate [this one](https://github.com/sunflowerit/odoo-modules/tree/8.0/project_task_markdown) as well?

File: web_widget_text_markdown/__manifest__.py (hunk 1 above)
Comment (KKamaa) [reply web_widget_text_markdown/__manifest__.py:7]: No its one commit named with "isort, prettier", This one I combined since when I made the first commit I had run pre-commit with my module added. But fixed it in [project_task_markdown](https://github.com/OCA/project/pull/828) and also learnt how to add a branch for travis cl in [doc](https://github.com/OCA/project/pull/828)

File: web_widget_text_markdown/__manifest__.py (hunk 2 above)
Comment (KKamaa) [reply web_widget_text_markdown/__manifest__.py]: Added

File: web_widget_text_markdown/README.rst (hunk 3 above)
Comment (KKamaa) [reply web_widget_text_markdown/README.rst]: followed the steps, also added Sunflower IT.

File: web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js (hunk 4 above)
Comment (KKamaa) [reply web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js:73]: I did the steps on migration after adding history, checked with pre-commit, nothing was found, then git add,  all the
changes into one, after noticing build failure,  I did:
`pre-commit run --all --show-diff-on-failure --verbose --color always` 
on my local and the command on maintainer-tools by studying previous pull requests.

File: web_widget_text_markdown/README.rst (hunk 3 above)
Comment (thomaspaulb) [reply web_widget_text_markdown/README.rst]: Almost correct: a contributor is always a person, so it should be kevin@sfit

File: web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js (hunk 4 above)
Comment (thomaspaulb) [reply web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js:73]: Hm, when I do the same, I get a different result:

```
cp -R web_widget_text_markdown /tmp  # backup your work
git reset --hard HEAD~2  # reset to before your work, but after adding history
pre-commit run --all --show-diff-on-failure --verbose --color always  # pre-commit on history 
git add . && git commit -m "[UPD] black, isort, prettier"   # add this
rm -Rf web_widget_text_markdown
cp -R /tmp/web_widget_text_markdown .  # warp files to after your work
pre-commit run --all --show-diff-on-failure --verbose --color always  # pre-commit, to be sure
git diff  # seeing lots of changes
git restore web_widget_text_markdown/i18n  # why were these removed?
git add . && git commit -m "[MIG] migration to 13.0"  # the commit that defines your migration activity
git push sunflowerit
```

Result is [here](https://github.com/sunflowerit/web/tree/13.0-mig-web-widget-text-markdown)

If you look at the [topmost commit](https://github.com/sunflowerit/web/commit/f4edca56fba47455be8b070cde42319c94cb1c54), now that is something I can easily review... and have left my comments there.

File: web_widget_text_markdown/README.rst (hunk 3 above)
Comment (KKamaa) [reply web_widget_text_markdown/README.rst]: fixed

File: web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js (hunk 4 above)
Comment (KKamaa) [reply web_widget_text_markdown/static/tests/js/web_widget_text_markdown.js:73]: well after a long while working out how to fix it, manage to understand basic concepts though complete fixture was not
added but clean up is ok. What I tried to do was:
```
git remote add sfit https://github.com/sunflowerit/web.git
git fetch sfit
git log sfit/13.0-web-widget-text-markdown
git cherry-pick f4edca5
git push origin head:13.0-web-widget-text-markdown
```

**Review by thomaspaulb — APPROVED** [review]

**Comment by thomaspaulb** [comment]
Note to other reviewers: dropdown code and translations are removed in favour of OCA module `web_drop_target`

**Review by thomaspaulb — APPROVED** [review]
Squash the last 3 commits.

**Review by dsolanki-initos — APPROVED** [review]
LGTM

File: web_widget_text_markdown/static/src/css/web_widget_text_markdown.css [hunk 5]
```diff
@@ -0,0 +1,7 @@
+.o_field_text_markdown blockquote {
+    background: #f9f9f9;
+    border-left: 10px solid #ccc;
+    margin: 1.5em 10px;
+    padding: 1em 10px 0.1em 10px;
+    quotes: "\201C""\201D""\2018""\2019";
+}
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/css/web_widget_text_markdown.css]: The 8.0 widget is showing me:

image

Latest 13.0 is showing me for the same:

image

My feeling is still that the 8.0 version is prettier, because of many small details that are actually all present in the [8.0 css](https://github.com/OCA/web/blob/8.0/web_widget_text_markdown/static/src/css/main.css#L128) which was not migrated to 13.0:

- [higher font sizes](https://github.com/OCA/web/blob/8.0/web_widget_text_markdown/static/src/css/main.css#L108)
- [paragraph margins](https://github.com/OCA/web/blob/8.0/web_widget_text_markdown/static/src/css/main.css#L135) that are the same above and below, and not only below
- [fancy codeblock styling](https://github.com/OCA/web/blob/8.0/web_widget_text_markdown/static/src/css/main.css#L294)
- [striped tables](https://github.com/OCA/web/blob/8.0/web_widget_text_markdown/static/src/css/main.css#L212)
- No weird margin at the top of a `<ul>` item (havent been able to figure out which CSS causes that)

Bottom line: why are we not just copypasting big chunks of the 8.0 css, and make it prettier without a lot of hard work?

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            LIBS_PATH + "bootstrap-markdown.min.css",
+            CUST_LIBS_PATH + "web_widget_text_markdown.css",
+        ],
+
+        _getValue: function() {
+            var $widget = this.attrs.widget;
+            var $type = this.field.type;
+            if ($type === "html" && $widget && $widget === "bootstrap_markdown") {
+                return this._getHtmlValue(this.$input.val());
+            }
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/js/web_widget_text_markdown.js:38]: Hm, this saves the converted HTML, but the next time you edit, you get HTML:

image

I'm OK with putting this feature here, but mention in the 'known issues' that if you use the widget on a HTML field, you lose the original markdown syntax.

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

File: web_widget_text_markdown/static/src/css/web_widget_text_markdown.css (hunk 5 above)
Comment (KKamaa) [reply web_widget_text_markdown/static/src/css/web_widget_text_markdown.css]: About this, we are using showdown js so it makes work easy but also when it comes to changing the bootstrap css its quite tricky. You can read on[ this](https://github.com/showdownjs/showdown/wiki/Add-default-classes-for-each-HTML-element),... it explains why this will be tricky but possible, I added a sample for testing.

Adding the css file directly doesn't affect every element unless i start adding important, or modify the _scss_ which might not be worth pursing, unless you do a sample and see what i mean.

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js (hunk 6 above)
Comment (KKamaa) [reply web_widget_text_markdown/static/src/js/web_widget_text_markdown.js:38]: this is now fixed, html is bidirectional -->  (html to markdown) and vice versa

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js [hunk 7]
```diff
@@ -42,8 +43,38 @@ odoo.define("web_widget_text_markdown.FieldTextMarkDown", function(require) {
...
+            // Showdown does not support this out of the box.
+            // eslint-disable-next-line max-len
+            // Ref: https://github.com/showdownjs/showdown/wiki/Add-default-classes-for-each-HTML-element
+            const classMap = {
+                table: "table table-striped",
+            };
+
+            const styleMap = {
+                h1: "font-size: 2.5em;",
+            };
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/js/web_widget_text_markdown.js]: I'm OK with using this method when classes are involved, for example `table-striped` is something useful and not easily added with CSS.

However if it comes to `h1` you will get code like this:

```
<h1 style="font-size: 2.5em;">Header</h1>
```

Using such inline styles is not a good practice IMO, your whole HTML code will get cluttered with inline style. For that I prefer just to use a CSS with

```
.o_field_text markdown { h1 { font-size: "2.5 em" }}
```

If it's not applying correctly, then perhaps some other style is taking precedence: in that case you can add `!important` to the end (but not if it's not necessary!)

Comment (KKamaa) [reply web_widget_text_markdown/static/src/js/web_widget_text_markdown.js]: @thomaspaulb As mentioned, the 13.0 bootstrap css file seems to override the v8.0 css changes i introduced.
see current commit for my implementation maybe i missed something.

But from my investigation the 13.0 bootstrap css uses directly the elements (h1, pre, table etc) so there aren't classes i can
override easily unless i also override directly which in this case affects Odoo. I am not sure if its cache, or hierarchy of adding assets maybe our css comes later after bootstrap. still investigating on the theories.

Also maybe we can settle for how 13.0 wrks without adding the v8.0 css changes since also it was not included in v11.0 which I began my port from. It might allow flexibility for users to choose their own theming styles, IMO.

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js (hunk 6 above)
Comment (KKamaa) [reply web_widget_text_markdown/static/src/js/web_widget_text_markdown.js:38]: Also added boolean under xml options.

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js (hunk 7 above)
Comment (thomaspaulb) [reply web_widget_text_markdown/static/src/js/web_widget_text_markdown.js]: @KKamaa Could it be the name of the class?

The widget has `.oe-*`:

image

Whereas the CSS has `.o-*`:

image

Comment (thomaspaulb) [reply web_widget_text_markdown/static/src/js/web_widget_text_markdown.js]: People are flexible to override this in their own way, but we need to provide a good default.

**Comment by hkapatel-initos** [comment]
Hello maintainers,
It would be nice if you are trying to merge it

**Comment by kmatthes** [comment]
@hkapatel-initos 
Here you asked to merge the PR, but the PR has no 2 approved reviews yet, has it?
Since KKamaa applyed some changes, maybe we can go in here with re-doing our review, either by @dsolanki-initos or you take over yourself. If we find nothing to claim about, we can kindly ask thomaspaulb to re-do his review, too. 
Ok?

**Comment by KKamaa** [comment]
@hkapatel-initos @kmatthes so the normal functionality still remains the changes I was doing was in reference to using the widget functionality with mail.template Html field. Though its up to reviewers on this point.

**Comment by hkapatel-initos** [comment]
@KKamaa work is complete or has any update?
please let us know after it is completed.

**Comment by hkapatel-initos** [comment]
@KKamaa please let us know after it is completed.

**Comment by KKamaa** [comment]
> @KKamaa please let us know after it is completed.

still pending review and update so will update this then it will be ready

**Comment by dsolanki-initos** [comment]
it takes a long time @KKamaa  let us know after it is completed.

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            LIBS_PATH + "bootstrap-markdown.js",
+            LIBS_PATH + "showdown.js",
+            LIBS_PATH + "showdown-footnotes.js",
+            LIBS_PATH + "showdown-table.js",
+            LIBS_PATH + "showdown-toc.js",
+        ],
+        cssLibs: [
+            LIBS_PATH + "bootstrap-markdown.min.css",
+            CUST_LIBS_PATH + "web_widget_text_markdown.css",
+        ],
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/js/web_widget_text_markdown.js]: Why arent these normal LESS files?

File: web_widget_text_markdown/static/src/css/web_widget_text_markdown.css [hunk 9]
```diff
@@ -0,0 +1,204 @@
+.o_field_text_markdown {
+    font-size: 15px;
+    line-height: 1.7;
+    overflow: hidden;
+}
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/css/web_widget_text_markdown.css]: I found the cause for the big amount of whitespace between two paragraphs; it's because Odoo adds

`white-space: pre-wrap`

to class `o_field_text`

In case the `o_field_text` is `o_field_text_markdown` it should just be `white-space: normal`, since the content is not only text but also HTML elements.

File: web_widget_text_markdown/static/src/css/web_widget_text_markdown.css (hunk 5 above)
Comment (thomaspaulb) [reply web_widget_text_markdown/static/src/css/web_widget_text_markdown.css]: Ok I get it. Let's declare this finished when:

- stuff we are adding is SCSS not CSS and gets compiled into the assets bundle like normal
- everything looks great and roughly the way on `project.task` like the way it was on 8.0 (so no loads of whitespace)

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

**Comment by thomaspaulb** [comment]
@dsolanki-initos If you want it finished please help the programmer to finish the open points by submitting pull requests. Otherwise it makes no sense to keep spamming. (cc @kmatthes)

**Comment by kmatthes** [comment]
Hey @dsolanki-initos , 
if I have a look at our last 3 comments, I can understand @thomaspaulb reaction. Let us discuss how to change our communication/process in such a case in our meeting today. (cc @alihashaam)

File: web_widget_text_markdown/static/src/css/web_widget_text_markdown.scss [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    & blockquote {
+        background: #f9f9f9;
+        border-left: 10px solid #ccc;
+        margin: 1.5em 10px;
+        padding: 1em 10px 0.1em 10px;
+        quotes: "\201C""\201D""\2018""\2019";
+    }
+
+    & pre {
+        background-color: #333 !important;
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/css/web_widget_text_markdown.scss:192]: color: white !important;

File: web_widget_text_markdown/static/src/js/web_widget_text_markdown.js [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        _getMarkdownValue: function(value) {
+            return this.shw_render_html.makeMarkdown(this._formatValue(value));
+        },
+        _getHtmlValue: function(value) {
+            return this.shw_render_html.makeHtml(this._formatValue(value));
+        },
+
+        _renderReadonly: function() {
+            this.$el.html(this._getHtmlValue(this.value));
+        },
```
Comment (thomaspaulb) [inline web_widget_text_markdown/static/src/js/web_widget_text_markdown.js:174]: For 'Implement Coupa' issue among others, it's sometimes losing italic/bold markup, and sometimes not.

Also, it's not always display existing italics in the text.

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

**Comment by dsolanki-initos** [comment]
@KKamaa will you please add suggested changes ?

**Comment by etobella** [comment]
@KKamaa Can you check the comments?

**Comment by thomaspaulb** [comment]
@etobella @dsolanki-initos We've hit a snag on this, the switch to using `showdown` turned out to be more involved than expected.

Sorry that it took long to respond.

I suggest that InitOS take the lead to open a new PR with a 'standard migration' which doesnt make any changes to the method, and I'll be happy to review that.

Meanwhile, we'll be opening a new 14.0 PR with the conversion from markdown to showdown somewhere in the next weeks.

Rationale of using showdown:

- The original module just allows using markdown in 'Text' widget fields, but not in 'Html' type fields.

- the 'showdown' module allows bidirectional conversion between HTML and text syntax, thus, the Markdown widget can operate on both Html and Text fields alike.

**Comment by thomaspaulb** [comment]
@KKamaa Let's close this.

**Comment by olgamarcocb** [comment]
There is an inconsistency with "Project" repository because there is a merged PR (#828) that depends on this module. 
Migration: #2136
